### PR TITLE
PLANNER-666: Execution server: Simplify response XML for OptaPlanner REST calls

### PR DESCRIPTION
**Pull request by mcimbora**
- Removed ServiceResponse wrapper from responses
- Reworked update state logic - instead of updating the state manually, use similar approach like UserTask (state transition via REST url). As to java client, introduced solvePlanningProblem & terminateSolverEarly.
- Polished java client interface

@mswiderski @ge0ffrey Could you please have a look? Thanks for your ideas.

### CODE REVIEW

**Comment by mcimbora** [comment]
@ge0ffrey I do believe solvePlanningProblem in java client describes the purpose of the method better than putStateSolving. Java client & rest endpoint are different worlds, difficult to find a tradeoff there. I don't think we should try to keep the naming consistent at all costs.

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/commands/optaplanner/SolvePlanningProblemCommand.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof SolvePlanningProblemCommand)) {
+            return false;
+        }
+
+        SolvePlanningProblemCommand that = (SolvePlanningProblemCommand) o;
+
+        if (containerId != null ? !containerId.equals(that.containerId) : that.containerId != null) {
```
Comment (ge0ffrey) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/commands/optaplanner/SolvePlanningProblemCommand.java]: Can't we use Objects.equal() here? Objects is a util class like Collections (the suffix s matters)

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java [hunk 2]
```diff
@@ -98,43 +99,44 @@ public void testSolverWithAsyncResponseHandler() throws Exception {
...
+
+        response = solverClient.getSolver(CONTAINER_1_ID, SOLVER_1_ID);
         assertThat(response);
         solver = responseCallback.get(SolverInstance.class);
         assertThat(solver.getSolverId()).isEqualTo(SOLVER_1_ID);
         assertThat(solver.getStatus()).isEqualTo(SolverInstance.SolverStatus.SOLVING);
 
-        response = solverClient.getSolverState(CONTAINER_1_ID, SOLVER_1_ID);
+        solverClient.terminateSolverEarly(CONTAINER_1_ID, SOLVER_1_ID);
+        responseCallback.get(Void.class);
```
Comment (ge0ffrey) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java]: What does responseCallback.get(Void.class) do?

**Comment by ge0ffrey** [comment]
We'll also need a PR on kie-docs to do any changes there if the docs are impacted (which I presume they are).

We 'll want to backport this to 7.0, before CR2 is released.

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/xstream/XStreamMarshaller.java [hunk 3]
```diff
@@ -159,11 +159,12 @@ protected void configureMarshaller( Set<Class<?>> classes, final ClassLoader cla
...
-        this.xstream.processAnnotations( GetBestSolutionCommand.class );
+        this.xstream.processAnnotations( GetSolverWithBestSolutionCommand.class );
         this.xstream.processAnnotations( GetSolversCommand.class );
-        this.xstream.processAnnotations( GetSolverStateCommand.class );
-        this.xstream.processAnnotations( UpdateSolverStateCommand.class );
-        
+        this.xstream.processAnnotations( GetSolverCommand.class );
+        this.xstream.processAnnotations( SolvePlanningProblemCommand.class );
+        this.xstream.processAnnotations( TerminateSolverEarlyCommand.class );
+
```
Comment (mswiderski) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/marshalling/xstream/XStreamMarshaller.java]: looks like missing imports for these new commands

File: kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/commands/optaplanner/SolvePlanningProblemCommand.java (hunk 1 above)
Comment (mcimbora) [inline kie-server-parent/kie-server-api/src/main/java/org/kie/server/api/commands/optaplanner/SolvePlanningProblemCommand.java]: I removed the equals method completely, other commands seem to rely on default object equality.

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java (hunk 2 above)
Comment (mcimbora) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java]: @ge0ffrey The callback queues results of latest service calls, this is to make sure the result is consumed (whatever the result type is).

**Comment by mcimbora** [comment]
Jenkins retest this.

**Comment by mcimbora** [comment]
Thanks for the review @ge0ffrey @mswiderski, this should be ready to go in. Indeed, let's hope someone superstitious doesn't revert the commit :-)

File: kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java (hunk 2 above)
Comment (ge0ffrey) [inline kie-server-parent/kie-server-tests/kie-server-integ-tests-optaplanner/src/test/java/org/kie/server/integrationtests/optaplanner/jms/OptaPlannerJmsResponseHandlerIntegrationTest.java]: Can we extract that one liner in a method in that unit test (or a parent unit test or a TestUtil) named something like makeSureAllResultsAreConsumed() - that way it's clear what the purpose is when reading the unit test (and the details of have that's done is less relevant).

**Comment by ge0ffrey** [comment]
@mcimbora do you want me to merge or wait for something like makeSureAllResultsAreConsumed()?
We'll want to backport to 7.0.x. (CR2 cut is Monday EOD.)

**Comment by mcimbora** [comment]
Jenkins retest this please

**Comment by mcimbora** [comment]
Jenkins retest this.

**Comment by manstis** [comment]
Merged to be consistent with 7.0.x PR https://github.com/kiegroup/droolsjbpm-integration/pull/923/

**Pull request merged**